openedx_tagging/core/tagging/rest_api/v1/serializers.py

```diff
@@ -30,11 +30,71 @@ class TaxonomyExportQueryParamsSerializer(serializers.Serializer):  # pylint: di
     output_format = serializers.RegexField(r"(?i)^(json|csv)$", allow_blank=False)
 
 
+class UserPermissionsSerializer(serializers.Serializer):  # pylint: disable=abstract-method
+    """
+    Serializer for the permissions on tagging instances granted to the current user.
+
+    Requires the current request to be passed into the serializer context, or all permissions will return False.
+    """
+    can_add = serializers.SerializerMethodField()
+    can_view = serializers.SerializerMethodField()
+    can_change = serializers.SerializerMethodField()
+    can_delete = serializers.SerializerMethodField()
+
+    def _check_permission(self, action, instance) -> bool:
+        """
+        Returns True if the current `request.user` may perform the given `action` on the `instance` object.
+
+        Uses the current request as passed into the serializer context.
+        """
+        assert action in ("add", "change", "view", "delete")
```

Comment:
nit:
The `add` action differs slightly from the others: we don't have the object instance to check against.

We will probably need a specific endpoint to check if the user can create/import a taxonomy. And we could use an attribute in the Taxonomy instance to check if a user can add tags to it.

Reply:
>  The add action differs slightly from the others: we don't have the object instance to check against.

Addressed with 8d619c5, thanks!

> We will probably need a specific endpoint to check if the user can create/import a taxonomy. And we could use an attribute in the Taxonomy instance to check if a user can add tags to it.

@bradenmacdonald suggested [doing it this way](https://github.com/openedx/modular-learning/issues/160#issuecomment-1881617717), so that we can see the model level permissions even if an empty list is returned.

I chose to implement it using a custom Paginator class, as the only way I could figure out how to inject this data alongside the [paginated response](https://github.com/encode/django-rest-framework/blob/047bec1288040b9f387e4977bb6cb3fd00d8ea40/rest_framework/pagination.py#L231-L237).

What do you both think of this approach? 23789cc

(To do: add model level permissions to views that use ObjectTagsByTaxonomySerializer, just not sure where yet.)

Reply:
@pomegranited I agree with @rpenido that it makes more sense to put only the "add" permission in the paginator and the view/change/delete permissions in each object. (Maybe we don't need `view` actually, because if you can't view it, it won't be included in the list. Right?)

@rpenido Good point about ObjectTags permissions depending on both the taxonomy and the object.

@rpenido We don't need to prevent the user from opening the drawer. Studio doesn't currently have a "read-only" view, so if you can see some library content or course content, you can edit it. But yes in the future if you can read tags but not edit them, you should still be able to open the drawer, just not make changes once it's opened.

Reply:
Thank you for your feedback @rpenido and @bradenmacdonald !

see 634168a for:

* only "can_add" returned at the top level.
* only "can_change" and "can_delete" returned for each object.

Will finish the rest of the feature and ping you for a full review.

Reply:
Nice @pomegranited , I think that's looking great.